### DISCUSSION TITLE
Add Paragraph alignment type : Thai-Distributed

### DISCUSSION THREAD

**Discussion by neenachcha** [dolanmiu/docx#1784] [Q&A] [answered]
Hi 

I have to export a word document in Thai and language, and make it looks like this.

IMAGE1
<img width="974" alt="Screen Shot 2565-11-18 at 13 32 14" src="https://user-images.githubusercontent.com/74401779/202636488-47599d92-ee53-4522-8050-dc520bc36bd4.png">

However, the document created from docx give me this. 

IMAGE2
<img width="987" alt="Screen Shot 2565-11-18 at 13 31 45" src="https://user-images.githubusercontent.com/74401779/202636383-dfcc0bd9-b4dc-49fd-9987-041feb58ed70.png">

This section supposes to have only 2 paragraphs, and both paragraphs must be evenly distributed (or justified).  I think the issue is about alignment type. I created document in IMAGE1 directly in word document, and use "Thai distributed alignment type". However, I created document in IMAGE2 from docx, and the alignment type shows "justify". I tried a few things but it doesn't improve the alignment.

Here are what I tried: 
1. I tried using AlignmentType.DISTRIBUTE
2. I tried adding compatibility

compatibility: {
      // version: 15, // not helpful,
      // doNotExpandShiftReturn: true, // not helpful,
      // wordPerfectJustification: true, // not helpful,
      // noTabStopForHangingIndent: true // not helpful,
 },

3. adding "Thai-distributed" into one of the option in alignment.d.ts

<img width="595" alt="Screen Shot 2565-11-18 at 13 35 59" src="https://user-images.githubusercontent.com/74401779/202637080-e6d83bd2-d9f6-42c1-9ef9-f18589f5c874.png">

I couldn't find other ways to make this work. 
It would be great if I could get some advice on how to make this work. 
And it would be great if docx could add "Thai-distributed" as one of the alignment type.

**Comment by dolanmiu** [comment]
I will look into this, sounds like a feature request

**Reply by neenachcha** [reply]
Thank you so much! I hope to hear from you soon.

**Reply by dolanmiu** [reply]
Can you upload a sample thai word file which I can use to test?

**Reply by dolanmiu** [reply]
I have added a PR to maybe fix this:

https://github.com/dolanmiu/docx/pull/1787

Hopefully this should fix it

Feel free to test it out for confirmation

**Reply by neenachcha** [reply]
[3.โรยัลกิด62_Notes.docx](https://github.com/dolanmiu/docx/files/10048388/3.62_Notes.docx)

Here is the sample of the thai word in docx with Thai-Distribution.

**Reply by neenachcha** [reply]
Thanks for taking your time to help me look into this. I tried to test by adding the "THAI_DISTRIBUTE = "thaiDistribute", into alignment.ts according to your PR. but it doesn't change alignment type in the exported docx. Did I miss anything?

**Reply by dolanmiu** [reply]
You are updating the type definitions, so it won't work `.d.ts`

You can use `docx` on master right now, and test, that should work I think

**Reply by dolanmiu** [reply]
I think my PR should add that feature into `docx`

Will be released for version `7.8.0`

**Reply by neenachcha** [reply]
Thanks for your reply. 
So I tested this. And Here is the result: 

<img width="994" alt="Screen Shot 2565-11-21 at 13 40 08" src="https://user-images.githubusercontent.com/74401779/202982197-8bc5dcc7-c9e8-4e05-b31d-2288504487b0.png">

The Alignment option is "Thai-distributed" as expected. However, the paragraphs aren't distributed properly. Please refer to IMAGE 1 - 4 for actual vs expected result below.

IMAGE1 (Actual - Export from docx)
<img width="608" alt="Screen Shot 2565-11-21 at 12 15 42" src="https://user-images.githubusercontent.com/74401779/202976000-7b49ea5f-e87e-4b10-9275-5ec4c5416812.png">

IMAGE2 (Expected)
<img width="607" alt="Screen Shot 2565-11-21 at 12 15 33" src="https://user-images.githubusercontent.com/74401779/202976265-753bb214-dd59-4c00-80ea-ccaca21c8709.png">

IMAGE3 (Actual - Export from docx)
<img width="595" alt="Screen Shot 2565-11-21 at 12 15 57" src="https://user-images.githubusercontent.com/74401779/202976108-e26c6ec5-926a-4923-a023-39989e4c37ba.png">

IMAGE4 (Expected)
<img width="606" alt="Screen Shot 2565-11-21 at 12 16 13" src="https://user-images.githubusercontent.com/74401779/202976441-8a617fb1-5526-4162-9c88-fe6062573bd3.png">

The docx document works as expected. If I paste Thai text inside docx, it will distributed properly. However, the text created from "new Paragraph" API shows difference result. 

`
new Paragraph({
        text:  "บริษัท บิสกิด จำกัด (บริษัทฯ) ได้จดทะเบียนจัดตั้งขึ้นเป็นบริษัทจำกัดตามประมวลกฎหมายแพ่งและพาณิชย์ของประเทศไทย เมื่อวันที่ 30 พฤษภาคม 2561 ทะเบียนนิติบุคคลเลขที่ 0845561005665",
        alignment:  AlignmentType.THAI_DISTRIBUTE,
        indent: {
          left: 350,
        }
      })
`

I think the text was split one time before it was pasted into docx? Do you have any idea on how do I fix this?

**Reply by dolanmiu** [reply]
I have added a thai distributed demo:

https://github.com/dolanmiu/docx/blob/master/demo/78-thai-distributed.ts

Seems to work as expected? Generated output:

![image](https://user-images.githubusercontent.com/2917613/203448601-8fcf89c1-b378-4ad1-8968-8ded9b741fbe.png)

**Reply by neenachcha** [reply]
This looks good. 

I couldn't fix it on my end. It looks like this. (Not equally distributed)
[บริษัท บิสกิด จำกัด_notes (2).docx](https://github.com/dolanmiu/docx/files/10103201/_notes.2.docx)

Would you mind help me test with different adjustment such as try adjusting the margin left, right or tried with different indentation and see if it still looks good? 

Maybe try this paragraph :

``
บริษัทฯ มีเงินสด 41,985.00 บาท และ 25,855.66 บาทตามลำดับ   เงินสดทั้งจำนวนอยู่ในความดูแลและรับผิดชอบของกรรมการ บริษัทฯบันทึกการรับชำระเงินและการจ่ายชำระเงินผ่านบัญชีเงินสดเพียงเท่านั้น ซึ่งอาจกระทบต่อความถูกต้องครบถ้วนของการบันทึกบัญชี ทั้งนี้ขึ้นอยู่กับระบบการควบคุมภายในของบริษัท
``

With this adjustment:
- font family : Angsana New
- Font size : 14
- Margin left : convertMillimetersToTwip(24)
- Margin right: convertMillimetersToTwip(24)
- Indent : convertMillimetersToTwip(6.4)

**Reply by dolanmiu** [reply]
It generated this:

![image](https://user-images.githubusercontent.com/2917613/204703666-64959ee6-e005-4f22-b8b5-f4b0f040ba39.png)

Can download it here:

[My Document.docx](https://github.com/dolanmiu/docx/files/10119223/My.Document.docx)

Is this looking correct? If so, I will release `7.8.0`

**Reply by neenachcha** [reply]
It looks correct in your photo. But It looks different on my side. This is what I downloaded in Window
![image](https://user-images.githubusercontent.com/74401779/204741201-909c7fd7-b7a9-4897-b0ae-ab819cfbffc8.png)

**Reply by neenachcha** [reply]
This one I downloaded from Macbook.
<img width="713" alt="Screen Shot 2565-11-30 at 15 08 06" src="https://user-images.githubusercontent.com/74401779/204741539-f1141c01-0dfc-40f5-9971-f6db29dbeacf.png">

It has too much space between letters in both window and macbook on my side.

**Reply by neenachcha** [reply]
Do you have any other adjustment? I use Microsoft365 Family for both Macbook and Window laptop.

**Reply by dolanmiu** [reply]
Very strange
No other adjustment no, I am using Word 2013. I did not set the font to `Angsana New`, just used the default font though. I do not have `Angsana New` installed on my computer

I uploaded this to Google Docs too, and it is fine for me:

![Screenshot 2022-11-30 at 14 49 21](https://user-images.githubusercontent.com/2917613/204827253-0d373020-5ec8-4c37-9a64-f86c158a245f.png)

I have created an example of how I created that doc:

https://github.com/dolanmiu/docx/blob/master/demo/79-thai-distributed.ts

**Reply by gozilla-paradise** [reply]
@dolanmiu 

I also got this problem too, the space between each character is weird.

<img width="1191" alt="image" src="https://github.com/dolanmiu/docx/assets/8213866/304fec27-dd9e-4ebf-b77b-4902091f1150">

I am using Word 16.76.1 for mac. The english language work just fine, the problem is only in Thai language. 

It works just fine on Google Doc but only weird in Microsoft Word.

I think the problem is how Thai Words are constructed. I have tried copy those weird spacing sentences in text editor and then paste back to Microsoft Word, and then they justified correctly.

Another workaround is to copy those sentences and the paste them again using "paste special with unformatted text option."

<img width="1144" alt="image" src="https://github.com/dolanmiu/docx/assets/8213866/aa9ede40-02d0-48c4-a3e0-1adde4c2b003">

**Reply by dolanmiu** [reply, marked as answer]
It is done, check out https://github.com/dolanmiu/docx/blob/master/demo/78-thai-distributed.ts
